docs/design/storagenode-graceful-exit-protocol.md

```diff
@@ -0,0 +1,215 @@
+The satellite verifies that the original piece hash matches the replacement piece hash, and verifies the order limits signature. On success, the satellite will update segment / pointer information. If verification fails, the satellite will send `ExitFailed`.
+
+![Transfer Sequence](./images/storagenode-graceful-exit-transfer-sequence.svg)
+
+#### Updating the Segment / Pointer
+
+When the upload has been verified the satellite will use `CompareAndSwap` to only switch the appropriate node without changing others. During the upload there may have been other things that happened:
+
+- Segment / pointer was reuploaded. We'll ignore piece transferred for the exit.
+- Segment / pointer pieces were changed, in that case we load the pointer and retry `CompareAndSwap`. When the storage node at that piece number has changed, we'll ignore the piece transferred for the exit, because one of them needs to be discarded.
+- Segment / pointer was deleted or is being deleted. We'll ignore the piece transferred for the exit.
+
+After changing the segment / pointer, we'll delete the piece from transfer queue, update progress, and send a `DeletePiece` to the exiting node.
+
+#### Handling transfer failure
+
+If a storage node report a failure to the satellite if it fails to upload to the replacement node, or the replacement node returns a bad piece hash. The satellite will track the failures and retry the transfer to another node.
+
+## Rationale
+
+We could have a separate initiate graceful exit RPC, however this would complicate things when satellite is unresponsive.
+
+## Implementation
+
+1. Add protobuf definitions.
+2. Implement verifying a transfer on the satellite.
+3. Implement transferring a single piece on storage node.
+4. Implement swapping node in segment.
+5. Implement non-async protocol for exiting.
+6. Implement async protocol for exiting.
+
+### Sketch
+
+```
+### storagenode ###
+
+for {
+    stream := dial satellite
+
+    for {
+        msg := stream.Recv(&msg)
+        if msg is NotReady {
+            go sleep a bit
+            and retry later
+        }
+
+        if msg is Completed {
+            update database about completion
+            delete all pieces
+            exit
+        }
+
+        if msg is TransferConfirmed {
+            update the progress table
+            delete the piece from disk
+            exit
+        }
+
+        if msg is TransferPiece {
+            // transfer multiple pieces in parallel, configurable
+            limiter.Go(func(){
+                result := try transfer msg.PieceID
+                stream.Send(result)
+                update progress table
+            })
+        }
+    }
+}
+
+### satellite ###
+
+if !initiated {
+    then add node to the gracefully exiting list
+    send NotReady
+    return
+}
+
+if !pieces collected {
+    send NotReady
+    return
+}
+
+inprogress pieces
+more pieces := true
+
+when storage node prematurely exits
```

Comment:
How does the satellite know if the storage node exited prematurely?

Reply:
It gets io.EOF when trying to read the next message.